packages/graphql/src/api-v6/queryIRFactory/ReadOperationFactory.ts

```diff
@@ -164,13 +166,28 @@ export class ReadOperationFactory {
         });
     }
 
-    private getPagination(connectionTreeArgs: GraphQLConnectionArgs): Pagination | undefined {
+    private getPagination(connectionTreeArgs: GraphQLConnectionArgs, entity: ConcreteEntity): Pagination | undefined {
         const firstArgument = connectionTreeArgs.first;
         const afterArgument = connectionTreeArgs.after ? cursorToOffset(connectionTreeArgs.after) : undefined;
         const hasPagination = firstArgument ?? afterArgument;
-        if (hasPagination) {
-            return new Pagination({ limit: firstArgument, skip: afterArgument });
+        const limitAnnotation = entity.annotations.limit;
+        if (hasPagination || limitAnnotation) {
+            const limit = this.calculatePaginationLimitArgument(firstArgument, limitAnnotation);
+            return new Pagination({ limit, skip: afterArgument });
+        }
+    }
+
+    private calculatePaginationLimitArgument(
+        firstArgument: Integer | undefined,
+        limitAnnotation: LimitAnnotation | undefined
+    ): number | undefined {
+        if (firstArgument && limitAnnotation?.max) {
+            return Math.min(firstArgument.toNumber(), limitAnnotation.max);
+        }
+        if (firstArgument && firstArgument.toNumber() >= (limitAnnotation?.max ?? Number.MAX_SAFE_INTEGER)) {
```

Comment:
These logic steps are a bit hard to read. Some variables to add context could help

Reply:
Let me document the steps!